### ISSUE TITLE
[FEATURE REQ] For the new SDKs, support serialization on Operation types.

### ISSUE THREAD

**Issue by jviau** [Azure/azure-sdk-for-net#13625]
**Library or service name.**
What library or service is this request related to? All new style SDKs.
 
**Is your feature request related to a problem? Please describe.**
What feature would you like to get added? What problem is it solving?

When writing a service that kicks off and waits on Azure operations, it is often useful to not sit on a thread, or even an async state machine, and poll an Azure operation all in one go. This is especially the case when using something like DurableTask or Azure Durable Functions, as there is a limited number of "Activities" you can have running at once. Instead, it is better to go back and forth from the queue performing single polls on an Azure operation until it is complete (for example, DurableTask can be backed by a service bus queue, so your "poll" operation is going back and forth from ServiceBus). It is also useful for error recovery, by being able to remember what operation you were waiting on, and resume waiting on it in the event of a failure. It would be very useful if the new SDKs `Operation` types support serialization, or some variant of it that is serializable. This way we can serialize that operation back and forth from the queue and ask it to update its state a single poll at a time.

This will require some design as these operations have internal service dependencies for performing the actual poll, http pipeline, and conversion to finals result. But if we were able to convert an operation to a minimal serialized object, then use some service from the SDKs to convert it back to a full Operation that would suffice. Type information for final result most likely can be left up to the code owner to know what type and SDK this operation is for, and they provide that information to the deserialization of the operation.

**Comment by tg-msft** [comment]
[We have a guideline that around this that says:](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-lro-subclass)

> ✔️ YOU MAY add additional APIs to subclasses of Operation<T>. For example, some subclasses add a constructor allowing to create an operation instance from a previously saved operation ID. Also, some subclasses are more granular states besides the IsCompleted and HasValue states that are present on the base class.

You can see an example of this at https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs#L932.  In this case, you could save the operation ID and blob's URI to recreate the `CopyFromUriOperation` to check the status or continue polling at a later time.

[We're not going to serialize enough information to recreate the entire client, though:](https://azure.github.io/azure-sdk/general_implementation.html#general-authimpl-no-persisting)

> ⛔️ DO NOT persist, cache, or reuse security credentials. Security credentials should be considered short lived to cover both security concerns and credential refresh situations.

**Comment by jviau** [comment]
@tg-msft thanks - I see that will be useful for the operations that do implement that. Looking at the new Azure Management SDKs, I don't see a way to save them and construct them later. In fact, the constructors they do have are internal. For example: [VirtualMachinesCreateOrUpdateOperation](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/compute/Azure.ResourceManager.Compute/src/Generated/VirtualMachinesCreateOrUpdateOperation.cs).

I understand not being able to serialize everything. I imagine it would be just the information to track this operation - as little as the URL(s) to poll on the operation, and then you would be expected to supply the client & credentials that would resume the polling on deserialization.

**Comment by tg-msft** [comment]
The new management SDKs are in preview and still working on implementing all of the guidelines.  I agree they should support being rehydrated though.

+@pakrym - Can we have the generator add Operation .ctor rehydration overloads automagically by default?

**Comment by jviau** [comment]
Thanks! One thing that would be nice to have, is if when it takes in a client, it takes in some common base class of the client. So we don't need to find the correct client to use per operation. This is possible with the current management SDKs. `AzureClientExtensions.GetLongRunningOperationResultAsync` is an extension method for an `IAzureClient`, allowing you to use anything that satisfies that interface to poll on operations from any management SDK.

**Comment by pakrym** [comment]
>+@pakrym - Can we have the generator add Operation .ctor rehydration overloads automagically by default?

Yes, https://github.com/Azure/autorest.csharp/issues/697

**Comment by tg-msft** [comment]
Assigning to @allenjzhang to consider a base class/interface for management plane clients.  Data plane auth is too varied to support a meaningful base class at this point.  I think the best long term answer will be serializing URIs and using `DefaultAzureCredential` to rehydrate clients when AAD is fully supported everywhere.

**Comment by jviau** [comment]
Base class for just management plan will be excellent for our use case.

**Comment by allenjzhang** [comment]
Thank you for your feedback. We will be considering this and get back to you in few weeks. Thanks.

AB#3998

**Comment by ArthurMa1978** [comment]
Tracked by https://github.com/Azure/autorest.csharp/issues/2158
